### ISSUE TITLE
incompatible types:, even if the types are the same

### ISSUE THREAD

**Issue by grainier** [ballerina-platform/ballerina-lang#17912]
**Description:**
$subject, when there are two modules involved.

**Steps to reproduce:**
1. Create a project and add two modules (i.e `core_module`, `sub_module`). 
2. Add the following in `core_module` source;
```ballerina
import ballerinax/java.jdbc;

jdbc:Client dbClient = new({
    url: "jdbc:mysql://localhost:3306/productsdb_300_wso2umuat?useTimezone=true&serverTimezone=UTC",
    username: "root",
    password: "root",
    poolOptions: { maximumPoolSize: 5 },
    dbOptions: { useSSL: false }
});

public function getClient() returns jdbc:Client {
    return dbClient;
}
```
3. Add following in `sub_module` source;
```ballerina
import ballerina/io;
import core_module;
import ballerinax/java.jdbc;

public function main() {
    jdbc:Client dbClient = core_module:getClient();
}
```
4.  Build the project, and at the compile time you'll get;
```
error: grainier/sub_module:0.1.0::main.bal:6:28: incompatible types: expected 'ballerinax/java.jdbc:0.0.0:Client', found 'ballerinax/java.jdbc:0.0.0:Client'
```

**Affected Versions:**
1.0.0-alpha

**OS, DB, other environment details and versions:**
Windows, Mac

**Related Issues (optional):**
https://stackoverflow.com/questions/57503996/ballerina-identical-types-are-incompatible

**Comment by grainier** [comment]
Reason for this seems to be; `jdbc:Client` has several private fields [1], and according to the spec [2]; when type checking, if there are private fields, and if it involves modules, then those types will be considered incompatible. To fix that, we might need to make those fields non-private.

1. https://github.com/ballerinax/jdbc/blob/master/jdbc-extension/src/main/ballerina/ballerinax/jdbc/endpoint.bal#L37
2. > if an object type descriptor T has private fields or methods, then it is not possible to define another object type descriptor that is a subtype of T, and
if an object type descriptor T has fields or methods with module-level visibility, then it is possible to define another object type descriptor that is a subtype of T only within the same module as T.

**Comment by pubudu91** [comment]
This seems to be due to a limitation in the isSameType() check
https://github.com/ballerina-platform/ballerina-lang/blob/daff974dc19a5275a3b3e1680c71b2924daa967f/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java#L1745
Will need to consider the module ID and the name of the type as well here, instead of just checking the references.

**Comment by MaryamZi** [comment]
This needs to be fixed on the language side, see https://github.com/ballerina-platform/ballerina-spec/issues/306.

**Comment by pubudu91** [comment]
Closing since this was fixed by #18045
